pkg/k8sutil/portforward.go

```diff
@@ -26,39 +27,90 @@ import (
 	"k8s.io/client-go/transport/spdy"
 )
 
-func IsPortAvailable(port int) bool {
+func IsPortAvailable(clientset kubernetes.Interface, port int) (bool, error) {
+	// kube-proxy stopped opening ports on the host for NodePort services in recent versions of kubernetes,
+	// so net.Listen won't be able to detect if the port is used by a NodePort service or not, and port forwarding might hang or redirect to the wrong component.
+	// https://github.com/kubernetes/kubernetes/pull/108496
+	// so we check if the port is used by a NodePort service.
+	services, err := clientset.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// TODO: make this work with minimal RBAC
+		if !kuberneteserrors.IsForbidden(err) {
+			return false, errors.Wrap(err, "failed to list services")
+		}
+	}
+
+	for _, service := range services.Items {
+		if service.Spec.Type != corev1.ServiceTypeNodePort {
+			continue
+		}
+		for _, p := range service.Spec.Ports {
+			if p.NodePort > 0 && int(p.NodePort) == port {
+				return false, nil
+			}
+		}
+	}
+
 	// portforward explicitly listens on localhost
 	host := net.JoinHostPort("127.0.0.1", strconv.Itoa(port))
 	server, err := net.Listen("tcp4", host)
 	if err != nil {
-		return false
+		return false, nil
 	}
 
 	server.Close()
-	return true
+	return true, nil
```

Comment:
Should we still do a net.Listen check at this point?  This function does not account for ports being in use by system processes that run outside of cluster.

Reply:
it accounts for ports being used where the kots commands are being run though, and that's still useful?

Reply:
yeah, nm, the code i had in mind is already in this function